nixos/modules/services/databases/clickhouse.nix

```diff
@@ -36,18 +36,18 @@ with lib;
 
     users.groups.clickhouse.gid = config.ids.gids.clickhouse;
 
+    systemd.tmpfiles.rules = [
+      "d ${stateDir} - clickhouse clickhouse - -"
+      "e ${confDir} - clickhouse clickhouse - -"
+    ];
```

Comment:
It seems scary to manage the database (which needs permanent storage) with systemd-tmpfiles (which services temporary files). According to https://github.com/systemd/systemd/pull/10802, the right fix is to prefix the commands in `ExecStartPre` with `+` (or `!`, or `!!`?).

Reply:
I'm going to ask @flokli to comment on this as he has made the point that using tmpfiles.d over mkdir allows modules to be more declarative.

Reply:
`systemd-tmpfiles` shouldn't be seen as a way to solely manage **temporary** files - the name is very misleading. As long as we don't provide an "age", they won't be cleaned up by systemd.

See https://github.com/systemd/systemd/commit/abcb67ceed56a722b9cf352fa27b4ba0076280c5 that updates the systemd's docs in that regard.

We should prefer using it to describe directory structures and files that need to be set up before a service can start, instead of calling commands in PreStart that massage the filesystem to the desired state.

In addition, from `man systemd-tmpfiles`:
> The age field only applies to lines starting with d, D, e, v, q, Q, C, x and X. If omitted or set to "-", no automatic clean-up is done

And on the `--clean` parameter for the `systemd-tmpfiles` command:
>  If this option is passed, all files and directories with an age parameter configured will be cleaned up.

So these tmpfiles aren't cleaned up, as long as we don't specify an age in the corresponding field.

Reply:
@orivej those guys are right - created directories are permanent, not temporary. The only downside is when someone disables `tmpfilesd` activation script, directories are no longer created before service starts. @oxij do you disable `tmpfilesd` in SLNOS? Or you also don't care about NixOS systemd services?

Reply:
The reason so many of these commits use tmpfiles.d is that the modules don't hardcode state and conf directories, but instead let the user specify them. Using StateDirectory, RuntimeDirectory, etc... should be preferred to tmpfiles.d when possible.

Now that I actually take a look at the code again I see that the author of that module *did* hardcode the path to live under /var/lib and /etc which is great! That means this commit *should* be fixed... but to use [StateDirectory and ConfigurationDirectory](https://www.freedesktop.org/software/systemd/man/systemd.exec.html) instead of tmpfiles.d, not mkdir though.

Keep in mind if stateDir and confDir were configurable tmpfiles.d still would be the correct way.

I'll go ahead and fix that...

Reply:
@orivej After reading some clickhouse [documentation](https://clickhouse.yandex/docs/en/getting_started/) and [source](https://github.com/yandex/ClickHouse/blob/master/debian/clickhouse-server.service) I've gone ahead and drafted https://github.com/NixOS/nixpkgs/pull/56265/commits/9a816f7c08976c81cca55f1489065b7958877afb.

Thoughts?

Reply:
@orivej ping

Does the new commit mentioned look good to you? Sorry about the confusion!

Reply:
Can the StateDirectory take an arbitrary absolute path as well?

Reply:
https://www.freedesktop.org/software/systemd/man/systemd.exec.html
> The specified directory names must be relative, and may not include "..".

No.